### ISSUE TITLE
xmlstream patch to get ready for TLS and SASL authenticators

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10701]
|<img alt="tofu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| tofu reported|
|-|-|
|Trac ID|trac#1046|
|Type|enhancement|
|Created|2005-05-20 21:29:28Z|

Attachments:

* [xmlstream.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/777/7776f8b9dd913610a9767b419953f947d654c1a7/0b1f2d11ea6408eef050651b1d6f21c2344bfb90.patch) (1840 bytes) - added by tofu on 2005-06-09 20:59:48Z - 
* [client.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/777/7776f8b9dd913610a9767b419953f947d654c1a7/918bab7a34a82ccdc56504d97a38bacd3f9551ca.patch) (15466 bytes) - added by tofu on 2005-06-09 21:00:31Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1046 1046
type__enhancement enhancement
reporter__tofu tofu
priority__highest highest
milestone__Words_0_5 Words-0.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1116624568000000 1116624568000000
changetime__1157360132000000 1157360132000000
version__2_0 2.0
owner__ 
cc__dizzyd cc__ralphm cc__tofu cc__hagna cc__TTimo cc__artdent cc__jadestorm cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="tofu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tofu commented|
|-|-|

```
#!html
<pre>
I just wanted to submit this patch to begin discussion on moving towards a TLS
and SASL authenticator. I will submit those patches later.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="tofu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tofu commented|
|-|-|

```
#!html
<pre>
New patch, the makes _reset() more generic.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm commented|
|-|-|

```
#!html
<pre>
You remove the separation between sending no version and sending version='1.0'.
I want this solved correctly and genericly for future versions. Also, your
handling of version identifiers is not correct, because it is not a float. A
version identifier is made up from two integers separated by a '.'. For example
'1.3' == '1.03' and also '1.12' > '1.9'. This is explained in RFC 3920.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="tofu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tofu commented|
|-|-|

```
#!html
<pre>
Doh, I will remove the patches and see if I can come up with something more
acceptable.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="tofu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tofu commented|
|-|-|

```
#!html
<pre>
New xmlstream patch. Let me know if this is ok.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="tofu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tofu commented|
|-|-|

```
#!html
<pre>
New client patch.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="TTimo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TTimo commented|
|-|-|

Is this patch enough to be able to connect to a TLS-only jabber server?
To which version of twisted does this apply? I tried against debian packages of 2.2.0 and getting rejects..

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm set status to `assigned`|
|-|-|

TLS support is already included in the current release and should work if you use `twisted.words.protocols.xmlstream`s directly. 

This issue is still open because there is no XMPP Authenticator yet that utilizes the TLS support. Also the SASL stuff is not in the trunk yet. The attached patches do not apply to the current code base. To finish this, work on [#11254](https://github.com/twisted/twisted/issues/11254) needs to be completed first.

**Comment by twisted-trac** [comment]
|<img alt="TTimo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TTimo commented|
|-|-|

Thanks for the clarification. I won't claim to understand all the technicalities, but I'd gladly help testing things if needed ( we run a TLS-only Jabber server with a mu-conference component )

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm commented|
|-|-|

Created [branches/xmpp-client-1046](branches/xmpp-client-1046) with a start of an implementation.

The end result should be an XMPP authenticator that tries TLS and uses SASL for authentication,
along with a factory to create the actual clients.

Currently done:

  * Implement the PLAIN and DIGEST-MD5 mechanisms
  * Connect using TLS
  * Authenticate using SASL

To be done:

  * Resource binding
  * Session establishment

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

Milestone Twisted-2.3 deleted

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm commented|
|-|-|

The resource binding and session establishment have now been done, and are known to be working. However, I am not very happy with how the steps are done.

What I think I would like is have `XmlStream`s have a list of _initializers_, that perform
part of the initializing of a session, in order. For regular Jabber clients, this list would be
something like 'ssl', 'sasl', 'bind', 'session'.

Initializers would be objects that keep their own state. For initiating entitities, they would check if the needed stream features have been provided by the server. Some steps may be configured to be optional.

My goal is that these initializers have an generic common interface that can be used to the stream initialization for receiving entities (servers), too.

Some of the code in the accompanying branch has been spun out and was addressed by ticket [#11241](https://github.com/twisted/twisted/issues/11241), that has since been committed. I merged forward and created a new branch: [branches/xmpp-client-1046-2.](branches/xmpp-client-1046-2.)

**Comment by twisted-trac** [comment]
|<img alt="jadestorm's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|jadestorm commented|
|-|-|

I've already spoken with ralphm about this, but just wanted to put this "in the record" so to speak.  Please conider adding SASL support for the component authenticator as well.  I must admit it's a fairly limited use, but Jabberd2 uses it and I make use of it in my python transports.

If it's at all helpful to look at it, I modified Tofu's code for these purposes:
http://svn.blathersource.org/filedetails.php?repname=pyicq-t&path=%2Ftrunk%2Fsrc%2Fsasl.py&rev=0&sc=0

(note that that's broken as of words 0.3.0, but it did the trick before then)

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm commented|
|-|-|

I committed a few things last week, including an implementation of a concept I've called _Initializers_. These should modularize the initialization steps like TLS, SASL, resource binding, etc. See [branches/xmpp-client-1046-2.](branches/xmpp-client-1046-2.)

Comments welcome.

**Comment by twisted-trac** [comment]
|<img alt="TTimo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TTimo commented|
|-|-|

Is there anything to test for mere mortals like me? Would I be able to use this branch to connect to an SSL-only Jabberd2 server?

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_22"></a>|@ralphm removed owner<br>@ralphm set status to `new`|
|-|-|

After working furiously on this for a while, I think it is ready to be merged into trunk. Please review.

@TTimo: this branch does not implement SSL, only XMPP 1.0 TLS and SASL. The new XMPPAuthenticator and associated XMPPClientFactory work similarly to BasicAuthenticator and basicClientFactory. Watch for the INIT_FAILED_EVENT event to catch everything that might go wrong in the stream initialization and inspect the failure object passed.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @radix|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_24"></a>|@radix set owner to @ralphm|
|-|-|

I think that almost the entire review will be about docstrings. Sorry :-)

For example, in client.py I see tons of new or modified methods, but not one docstring. 

BTW, the NEWS file conflicts when merged to trunk. Not a big deal.

sasl.py,
 * while to_b64 is extremely trivial, its necessity is unclear, so a docstring would be appreciated there.
 * Same with get_mechanisms. 
 * from_b64 almost certainly shouldn't exist.
 * Everything in this module needs a docstring.

sasl_mechanisms.py is bereft of docstrings.

xmlstream.py:
 * useTls should be named useTLS, as (somewhat) per the coding standard.
 * ConnectAuthenticator.initializeStream seems to refer to some identifiers and methods that don't actually exist or that it doesn't actually call.

Tests:
 * We have some very good examples of  docstrings here, but docstrings need to be more thoroughly applied to tests (This isn't actually part of the coding standard yet, but it has become convention as of a few months ago that all new test methods need docstrings clarifying their general intent).

In general on docstrings: I don't care if you don't document totally private methods (as long as they have a prefix of _), but I do think it's important to document the implementation of interfaces: in these cases, the docstring should try to describe what makes *this* implementation different from the others. I'm also generally happy with the docstrings you have written; we just need more of the same :-)

Sorry for being so repetitive here, but I think it would help me give you a better review of the code if I had more docstrings to read, so please add them and I'll look at the code again.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_25"></a>|@ralphm set owner to @radix|
|-|-|

I removed the reference to useTls in component.py, it wasn't present in xmlstream.py anymore. Also added numerous docstrings. See if you like this better and please review again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun set owner to @ralphm|
|-|-|

Why isn't TLSInitiatingInitializer a BaseInitiatingInitializer subclass?  If it were, it could avoid duplicating the three-branch conditional in initialize() and just put logic into start.

TLSInitiatingInitializer.onProceed can use ```ssl.CertificateOptions()``` instead of making a ```ClientContextFactory``` and settings its method to ```SSL.TLSv1_METHOD```.

In epytext markup, you can't link to builtin types (ie, L{int}).  So, for example, XmlStream.version's type should be documented as something like "C{(int, int)}" or "two-tuple of C{int}".

Likewise, L{str}, L{bool}, etc, should use C{}.

XmlStream.initiating is documented with "@jvar"

XmlStream.initializers' type says it is a list of objects which implement a particular interface.  I think the wording here is intended to be "provides".

```python
class X: implements(Y)
a = X()
```

In the above, ```X``` implements ```Y```.  ```a``` provides ```Y```.

The XmlStream.initializers class attribute should probably be () instead of [] to avoid the accident of mutating the class attribute (which it seems to me would be an unfortunate event).

Still some classes/methods without docstrings:
 * FeatureNotAdvertized
 * IQAuthInitializer
 * CheckVersionInitializer
 * BindInitializer
 * SessionInitializer
 * XMPPClientFactory
 * XMPPAuthenticator

New code with no test coverage (determined with trial --coverage):
 * IQAuthInitializer.initialize
 * positive path in BindInitializer.onBind
 * SessionInitializer.start
 * XMPPClientFactory
 * XMPPAuthenticator.__init__
 * XMPPAuthenticator.associateWithStream
 * positive path in branch at the end of DigestMD5.__init__
 * The rest of DigestMD5 (note also that there are at least two other DigestMD5 implementations in Twisted - we are really, really, really getting to the point where some refactoring needs to happen)
 * BaseFeatureInitiatingInitializer.initialize
 * TLSInitiatingInitializer.initialize

Another conflict has cropped up in test_jabberxmlstream.py, so I just resolved the conflicts and merged the branch forward, since I need to resolve them anyway to run the test suite.  Code is now in xmpp-client-1046-3

Generally, branch is looking pretty good.  I'm not exactly sure how the new code gets used, though.  It might be nice to update the jabber_client.py example to use the SASL and TLS features?

**Comment by twisted-trac** [comment]
|<img alt="TTimo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_27"></a>|TTimo commented|
|-|-|

If sample code gets updated, you have a tester right there. I still have no idea how to use the new stuff.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_28"></a>|@ralphm set status to `assigned`|
|-|-|

I added a number of docstrings and fixed the mistakes in the existing ones. Also took the suggestion of making `TLSInitializer` derive from `BaseFeatureInitiatingInitializer`.

About epytext and builtins: I have not read about such limitation in the epytext documentation. Epydoc did process them to link to the types' interface description, just like any other object. Every time this comes up in #twisted, there seems to be no consistent stance on this. If there is the desire to remove all such links (that currently harm no one), it should be raised in a seperate ticket that also updates the coding standard policy to match.

The `ssl.CertificationOptions` appears to be written for the server side of the connection (the docstrings mention this), whereas `TLSInitializer` implements the client side. I would like to be able to use client certificates and verify server certificates against a local registry of CAs. That is the reverse of what `ssl.CertificationOptions` now implements, but maybe it can be generalized to accomodate both parties?

The `XmlStream.initializers` class attribute is always an ordered list that is stepped through
by popping the first item repeatedly. I can imagine situations where the list is modified by an initializer, for example having an initializer that uses different authentication initializers depending on the stream version. I haven't actually tried this yet, but in theory it would insert the selected authenticator after the first (that is currently being processed) and then return successfully.

Thanks for the conflict resolution.

Continuing work on the tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun commented|
|-|-|

Replying to [ralphm](#note_28):
> About epytext and builtins: I have not read about such limitation in the epytext documentation. Epydoc did process them to link to the types' interface description, just like any other object. Every time this comes up in #twisted, there seems to be no consistent stance on this. If there is the desire to remove all such links (that currently harm no one), it should be raised in a seperate ticket that also updates the coding standard policy to match.

I don't think pydoctor likes these.  

> 
> The `ssl.CertificationOptions` appears to be written for the server side of the connection (the docstrings mention this), whereas `TLSInitializer` implements the client side. I would like to be able to use client certificates and verify server certificates against a local registry of CAs. That is the reverse of what `ssl.CertificationOptions` now implements, but maybe it can be generalized to accomodate both parties?

Re-capping what we talked about on IRC, CertificateOptions is actually usable for both the client and server sides of the conversation, despite its documentation.  So there shouldn't be any problem here.

> 
> The `XmlStream.initializers` class attribute is always an ordered list that is stepped through
> by popping the first item repeatedly. I can imagine situations where the list is modified by an initializer, for example having an initializer that uses different authentication initializers depending on the stream version. I haven't actually tried this yet, but in theory it would insert the selected authenticator after the first (that is currently being processed) and then return successfully.

So it sounds like this is _intended_ to be mutable global state?  Euurgg :/  Generally, I think this is a bad idea.  There's no reason to have an API for forcing behavior onto _every_ XmlStream.  Conch did/does something similar, and it is/was a disaster.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_30"></a>|@ralphm commented|
|-|-|

I'll look into the `CertificateOptions` thing now.

I don't understand the last comment. Or do you mean I should not initialize it as a class variable but as an instance variable?

Added full tests for BindInitializer, SessionInitializer, IQAuthInitializer, TLSInitializer and BaseFeatureInitiatingInitializer.

I'd would very much like to have on generic SASL implementation. I don't know which should be used, mine are protocol agnostic. As you mentioned the DigestMD5 mechanism lacks tests, currently.

I find the XMPPAuthenticator and XMPPClientFactory fairly straightforward. Not sure I need to add tests for them.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_31"></a>|@exarkun commented|
|-|-|

Replying to [ralphm](#note_30):
> 
> I don't understand the last comment. Or do you mean I should not initialize it as a class variable but as an instance variable?

Yep, pretty much.

> 
> Added full tests for BindInitializer, SessionInitializer, IQAuthInitializer, TLSInitializer and BaseFeatureInitiatingInitializer.
> 

Hooray.

> I'd would very much like to have on generic SASL implementation. I don't know which should be used, mine are protocol agnostic. As you mentioned the DigestMD5 mechanism lacks tests, currently.
> 

Created ticket [#2015](https://github.com/twisted/twisted/issues/2015) for this.

> I find the XMPPAuthenticator and XMPPClientFactory fairly straightforward. Not sure I need to add tests for them.

XMPPAuthenticator is doing a lot.  It is providing at least 5 features to an XmlStream.  I think it should definitely have tests.

XMPPClientFactory is pretty simple, but it should still be tested.  Ideally, there would be some unit tests dedicated solely to it, but I would settle for it at least being exercised by some other test.

Code that doesn't even get run by the test suite must be pointless code, since all features should have unit tests.  If code is used by no feature, why does it exist at all?

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_32"></a>|@ralphm commented|
|-|-|

I added a test case both `XMPPAuthenticator` and `XMPPClientFactory`, and also now use `ssl.CertificateOptions`. I can imagine the TLS growing support for certificates soon, but I want to do that in another branch. That might also complicate the `XMPPClientFactory` and `XMPPAuthenticator` a bit, as the client certificate must be supplied.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_33"></a>|@ralphm set owner to @exarkun<br>@ralphm set status to `new`|
|-|-|

Added Twisted/doc/words/examples/xmpp_client.py.

I have the intention of extracting the common bits here and put them in t.w.p.j.client. But that can be done in another branch.

I suppose I'm done. Please review again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_34"></a>|@exarkun set owner to @ralphm|
|-|-|


Some minor stuff:

 * codes_to_conditions should probably be CODES_TO_CONDITIONS

 * in twisted.words.test.test_jabberxmlstream.XmlStreamTest.setUp, it would probably be best to use ```xs.makeConnection(transport)``` rather than ```xs.transport = transport```.  This will have the same effect, but follow the actual codepath used outside the test suite.  One difference is that this will also cause .connectionMade() to be called on the xmlstream.  I notice that some test methods do this explicitly and some don't.  I'm not sure if this is intentional, but I'd guess it is not the case that the intent of the tests is to assert that, eg, ```sendHeader``` works before a connection has been established.

 * TLSInitiatingInitializerTest should probably save the value of xmlstream.ssl.supported in setUp and restore it in tearDown (and ```savedSupport```, not ```saved_support```, please :).  This removes some LOC from each test method, but also properly restores the saved value in case the test method fails, which will currently skip the code which restores the saved value. (Nice tests here, btw)

 * TLSInitiatingInitializerTest.testFailed is missing a docstring, as are BaseFeatureInitiatingInitializerTest's test methods.

 * SASLInitiatingInitializerTest.testOnFailure also needs a docstring, as do the test methods in test_jabbersaslmechanisms

 * XMPPAuthenticatorTest.testBasic needs a docstring

There are still a bunch of docstrings that use the single-leading space style.  I'm fine with leaving those as-is for this branch, if you make another ticket for reformatting them.

Other than that, I think everything in this branch looks pretty good now.  The only thing I have a concern about is the mouthful of a name presented by the "InitiatingInitializer" classes.  I don't think this necessarily needs to change in this branch, but I think it should at some point (I notice the example doesn't need to name any of these classes explicitly, but if more advanced code is going to start using them, changing it now might be for the best).  As I understand it, these will have counterparts when the server code gets added, ReceivingInitializer or something like that?  Off the top of my head, something like InitializerClient/InitializerServer makes sense.  (I realize the TCP client is not necessarily the party which initiates initialization, but there are other examples in Twisted of disassociating the notion of tcp client/server from the higher-level protocol's client/server.)

Thanks for sticking with this through all the back-and-forth :)

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_35"></a>|@ralphm set owner to @exarkun|
|-|-|

I think I addressed all the docstrings and suggestions to improve the
tests.

About the naming of the initializers. The name 'initiating' comes from the terminology used in the XMPP specification (RFC 3920). An XML stream has two parties: the initiating entity and the receiving entity. In all cases is the initiating entity the TCP client.

Then, there are several types of XML streams, differentiated by the default namespace used. You have client-to-server streams (between an XMPP client and an XMPP server), server-to-server streams (between two XMPP servers) and component-to-server streams (to connect an external server component to its server). So, I think using client/server terminology to point out the different parties to a stream might be very confusing.

Any ideas about improving the terminology is welcome, of course.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_36"></a>|@exarkun set owner to @ralphm|
|-|-|

This is looking pretty good now.  I did a whitespace cleanup commit (some tabs got in in one of your recent commits and some docstrings still looked wrong).  I think the branch is good to merge now.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_37"></a>|@ralphm set status to `closed`|
|-|-|

(In [18041]) Implement basic XMPP 1.0 client support.

Author: ralphm
Reviewer: exarkun
Fixes [#10701](https://github.com/twisted/twisted/issues/10701)

This adds SASL authentication (PLAIN and DIGEST-MD5), resource binding and
session establishment, reworks TLS support and provides a new factory for
XMPP client connections.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
